mirrord-layer/src/files.rs

```diff
@@ -0,0 +1,194 @@
+        // sudden. We have to ignore this here as `node` will look for it outside the cwd (it
+        // searches up the directory, until it finds).
+        r".*/package.json",
+        &current_dir.to_string_lossy(),
+    ])
+    .unwrap();
+
+    set
+});
+
+static OPEN_FILES: SyncLazy<Mutex<HashSet<RawFd>>> =
+    SyncLazy::new(|| Mutex::new(HashSet::with_capacity(4)));
+
+#[derive(Debug, Clone)]
+pub(crate) struct ReadFile {
+    bytes: Vec<u8>,
+    read_amount: usize,
+}
+
+// TODO(alex) [high] 2022-05-19: Hook file `seek`, and `write`.
+
+/// Blocking wrapper around `libc::open` call.
+///
+/// It's bypassed when trying to load system files, and files from the current working directory
+/// (which is different anyways when running in `-agent` context).
+///
+/// When called for a valid file, it'll block, send a `ClientMessage::OpenFileRequest` to be handled
+/// by `mirrord-agent` (`handle_peer_message` function), and wait until it receives a
+/// `DaemonMessage::FileOpenResponse`.
+pub(crate) fn open(path: PathBuf) -> Result<RawFd, LayerError> {
+    debug!("open -> trying to open valid file {path:?}.");
+
+    let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+    let (file_channel_tx, file_channel_rx) = oneshot::channel::<OpenFileResponse>();
+
+    let requesting_file = OpenFileHook {
+        path,
+        file_channel_tx,
+    };
+
+    sender.blocking_send(HookMessage::OpenFileHook(requesting_file))?;
+
+    debug!("Blocking on file open operation!");
+    let open_file = file_channel_rx.blocking_recv()?;
+    debug!("After `block_on` call, we have an open file {open_file:#?}!");
+
+    OPEN_FILES.lock().unwrap().insert(open_file.fd);
+
+    Ok(open_file.fd)
+}
+
+/// Blocking wrapper around `libc::read` call.
+///
+/// Bypassed when trying to load system files, and files from the current working directory, see
+/// `open`.
+pub(crate) fn read(fd: RawFd, read_amount: usize) -> Result<ReadFile, LayerError> {
+    debug!("read -> trying to read valid file {fd:?}.");
+
+    let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+    let (file_channel_tx, file_channel_rx) = oneshot::channel::<ReadFileResponse>();
+
+    let read_file = ReadFileHook {
+        fd,
+        buffer_size: read_amount,
+        file_channel_tx,
+    };
+
+    sender.blocking_send(HookMessage::ReadFileHook(read_file))?;
+
+    let read_file_response = file_channel_rx.blocking_recv()?;
+
+    let read_file = ReadFile {
+        bytes: read_file_response.bytes,
+        read_amount: read_file_response.read_amount,
+    };
+
+    Ok(read_file)
+}
+
+/// NOTE(alex): libc also has an `open64` function. Both functions point to the same address, so
+/// trying to intercept them all will result in an `InterceptorAlreadyReplaced` error.
+pub(super) unsafe extern "C" fn open_detour(raw_path: *const c_char, oflag: c_int) -> RawFd {
+    let path: PathBuf = CStr::from_ptr(raw_path)
```

Comment:
I'd move this code to the safe function, as we want as less unsafe as possible (same for other detour functions)